Calling 'registerOutputs' twice leads to a crash.

This happened to me due a coding bug in a ComponentResource hierarchy.  I accidently had both a base-class and a derived-class **both** call `.registerOutputs`.  This is certainly a bug in the hierarchy.  However, we should still not crash the engine, and should instead report some reasonable error to the user.

The crash stack is:

```go
panic: fatal: An assertion has failed: cannot complete a resource 'urn:pulumi:cyservtest::containers-fargate::awsinfra:x:ecs:FargateService::fargate-cached-nginx' whose registration isn't pending

goroutine 82 [running]:
github.com/pulumi/pulumi/pkg/util/contract.failfast(0xc00361c000, 0xb5)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/util/contract/failfast.go:23 +0xe9
github.com/pulumi/pulumi/pkg/util/contract.Assertf(0xe03400, 0xef704e, 0x40, 0xc001709b48, 0x1, 0x1)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/util/contract/assert.go:33 +0x126
github.com/pulumi/pulumi/pkg/resource/deploy.(*stepExecutor).ExecuteRegisterResourceOutputs(0xc000616500, 0x7fa1d0b17770, 0xc000d09420)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_executor.go:153 +0x16b
github.com/pulumi/pulumi/pkg/resource/deploy.(*planExecutor).handleSingleEvent(0xc000d2f360, 0xff84e0, 0xc000d09420, 0xc001709cd8)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan_executor.go:217 +0x2ea
github.com/pulumi/pulumi/pkg/resource/deploy.(*planExecutor).Execute.func3(0xc000cdd0e0, 0xc000d2f360, 0xc000c5fc00, 0x10054a0, 0xc0005f1640, 0x10054a0, 0xc0005f1500, 0x10000, 0x1, 0xc000616500)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan_executor.go:171 +0x280
github.com/pulumi/pulumi/pkg/resource/deploy.(*planExecutor).Execute(0xc000d2f360, 0x10054a0, 0xc0005f1500, 0x1001a20, 0xc000602870, 0x7fffffff, 0x10000, 0xc0005d3201, 0x0, 0x0)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan_executor.go:187 +0x4c4
github.com/pulumi/pulumi/pkg/resource/deploy.(*Plan).Execute(0xc00171b3e0, 0x10054a0, 0xc0005f1500, 0x1001a20, 0xc000602870, 0x7fffffff, 0x10000, 0xc000da5001, 0x1000, 0x1000)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan.go:295 +0x87
github.com/pulumi/pulumi/pkg/engine.(*planResult).Walk.func1(0x1001a20, 0xc000602870, 0xc0006027e0, 0x10054a0, 0xc0005f1500, 0x1, 0xc000c5fb80, 0xc000cdc9c0)
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/engine/plan.go:216 +0xc9
created by github.com/pulumi/pulumi/pkg/engine.(*planResult).Walk
	/home/cyrusn/go/src/github.com/pulumi/pulumi/pkg/engine/plan.go:208 +0x114
```

--- a/changelog/pending/20230927--engine--calling-registerresourceoutputs-twice-no-longer-panics-and-returns-an-error-instead.yaml
+++ b/changelog/pending/20230927--engine--calling-registerresourceoutputs-twice-no-longer-panics-and-returns-an-error-instead.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: engine
+  description: Calling RegisterResourceOutputs twice no longer panics and returns an error instead.

--- a/pkg/resource/deploy/step_executor.go
+++ b/pkg/resource/deploy/step_executor.go
@@ -149,7 +149,9 @@ func (se *stepExecutor) ExecuteRegisterResourceOutputs(e RegisterResourceOutputs
 	// Look up the final state in the pending registration list.
 	urn := e.URN()
 	value, has := se.pendingNews.Load(urn)
-	contract.Assertf(has, "cannot complete a resource '%v' whose registration isn't pending", urn)
+	if !has {
+		return fmt.Errorf("cannot complete a resource '%v' whose registration isn't pending", urn)
+	}
 	reg := value.(Step)
 	contract.Assertf(reg != nil, "expected a non-nil resource step ('%v')", urn)
 	se.pendingNews.Delete(urn)

--- a/pkg/resource/deploy/step_executor_test.go
+++ b/pkg/resource/deploy/step_executor_test.go
@@ -1,0 +1,53 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRegisterResourceErrorsOnMissingPendingNew(t *testing.T) {
+	t.Parallel()
+
+	se := &stepExecutor{
+		pendingNews: sync.Map{},
+	}
+	urn := resource.URN("urn:pulumi:stack::project::my:example:Foo::foo")
+	err := se.ExecuteRegisterResourceOutputs(&mockRegisterResourceOutputsEvent{
+		urn: urn,
+	})
+	// Should error, but not panic since the resource is being registered twice.
+	assert.Error(t, err)
+}
+
+type mockRegisterResourceOutputsEvent struct {
+	urn resource.URN
+}
+
+var _ = RegisterResourceOutputsEvent((*mockRegisterResourceOutputsEvent)(nil))
+
+func (e *mockRegisterResourceOutputsEvent) event() {}
+
+func (e *mockRegisterResourceOutputsEvent) URN() resource.URN { return e.urn }
+
+func (e *mockRegisterResourceOutputsEvent) Outputs() resource.PropertyMap {
+	return resource.PropertyMap{}
+}
+
+func (e *mockRegisterResourceOutputsEvent) Done() {}

